pkg/reconcilermanager/controllers/reconciler_base.go

```diff
@@ -639,16 +640,69 @@ func (r *reconcilerBase) setupOrTeardown(ctx context.Context, syncObj client.Obj
+			r.logger(ctx).Error(err, "The reconciler-manager finalizer removal failed")
 			return err
 		}
-		r.logger(ctx).Info("Finalizer removal successful")
+		r.logger(ctx).Info("The reconciler-manager finalizer removal successful")
 	}
 
 	return nil
 }
 
+// removeReconcilerFinalizer removes the 'configsync.gke.io/reconciler'
+// finalizer from the RepoSync object when the RepoSync Namespace is terminating.
+//
+// Note: it ONLY removes the finalizer from RepoSync, but NOT RootSync. If
+// RootSync is being deleted but the finalizer still exists, we'll wait for the
+// reconciler finalizer to complete. It is possible RootSync is stuck in
+// deletion because of the finalizer (e.g. root-reconciler is either deleted, or
+// unable to remove the finalizer), but we decided to leave it as a known issue
+// and brainstorm a better solution later.
+//
+// The reconciler adds the finalizer when deletion propagation is set to
+// `Foreground`, and removes it when RSync has a deletionTimestamp.
+// When the RepoSync's Namespace is terminating, the RoleBinding for the
+// namespace reconciler could be deleted and fail to be re-created.
+// Therefore, the namespace reconciler would lack the necessary permissions to
+// remove the reconciler finalizer from the RepoSync object, leading to
+// RepoSync being stuck in deletion. This function allows the reconciler-manager
+// to delete the reconciler finalizer in this case.
+// It returns whether the reconciler finalizer has been removed.
+// If removed, the controller can proceed to tear down other resources.
+func (r *reconcilerBase) removeReconcilerFinalizer(ctx context.Context, syncObj client.Object) (bool, error) {
+	if syncObj.GetObjectKind().GroupVersionKind().Kind == configsync.RootSyncKind {
+		r.logger(ctx).Info("Waiting for the root reconciler finalizer to finish")
+		return false, nil
+	}
+
+	syncObjNamespace := &corev1.Namespace{}
+	syncObjNamespace.Name = syncObj.GetNamespace()
+	key := client.ObjectKeyFromObject(syncObjNamespace)
+	if err := r.client.Get(ctx, key, syncObjNamespace); err != nil {
+		r.logger(ctx).Error(err, "Get sync object Namespace failed")
+		return false, err
+	}
+	if syncObjNamespace.Status.Phase == corev1.NamespaceTerminating {
+		if !controllerutil.RemoveFinalizer(syncObj, metadata.ReconcilerFinalizer) {
+			r.logger(ctx).Info("The reconciler finalizer is not found")
+			return true, nil
+		}
+		if err := r.client.Update(ctx, syncObj); err != nil {
+			err = status.APIServerError(err,
+				fmt.Sprintf("failed to update %s to remove the reconciler finalizer", r.syncKind))
+			r.logger(ctx).Error(err, "The reconciler finalizer removal failed")
+			return false, err
+		}
+		r.logger(ctx).Info("The reconciler finalizer removal successful")
+		return true, nil
+	}
+
+	// RepoSync's Namespace is active, then wait for the reconciler finalizer to complete.
+	r.logger(ctx).Info("Waiting for the namespace reconciler finalizer to finish")
```

Comment:
nit: We don't need to remove finalizer if the namespace is active, so it's fine not to log it. Suggest a slight change to make the code more clean:

```go
if syncObjNamespace.Status.Phase != corev1.NamespaceTerminating {
  return false, nil
}

// <move L687-698 here>
```

Reply:
done